### DISCUSSION TITLE
2차 강의에서 로봇 조종 관련하여 질문 드립니다.

### DISCUSSION THREAD

**Discussion by Jeoungji** [Road-Balance/gcamp_ros2_basic#48] [Q&A]
2차 강의에서 나온 airobo_gazebo에서 로봇은 나오나 주행이 되지 않습니다.
![오류13](https://user-images.githubusercontent.com/109124955/179003617-5109d269-ebaf-4fb6-8029-eb46a19a646b.PNG)

ros2 launch airobo_gazebo gazebo_world.launch.py
를 실행하였을 때 나온 것들입니다.

(ROS 2 foxy) dnw0406@DESKTOP-MD97PB4:~/ros2_ws$
ros2 launch airobo_gazebo gazebo_world.launch.py
[INFO] [launch]: All log files can be found below /home/dnw0406/.ros/log/2022-07-14-22-29-30-630676-DESKTOP-MD97PB4-1452
[INFO] [launch]: Default logging verbosity is set to INFO
 If it's your 1st time to download Gazebo model on your computer, it may take few minutes to finish.
[INFO] [gzserver-1]: process started with pid [1454]
[INFO] [gzclient   -2]: process started with pid [1457]
[INFO] [robot_state_publisher-3]: process started with pid [1460]
[INFO] [joint_state_publisher-4]: process started with pid [1462]
[INFO] [spawn_entity.py-5]: process started with pid [1464]
[robot_state_publisher-3] Parsing robot urdf xml string.
[robot_state_publisher-3] Link chassis had 9 children
[robot_state_publisher-3] Link back_left_wheel had 0 children
[robot_state_publisher-3] Link back_right_wheel had 0 children
[robot_state_publisher-3] Link camera had 0 children
[robot_state_publisher-3] Link front_left_wheel had 0 children
[robot_state_publisher-3] Link front_right_wheel had 0 children
[robot_state_publisher-3] Link hokuyo had 0 children
[robot_state_publisher-3] Link imu_link had 0 children
[robot_state_publisher-3] Link left_track had 0 children
[robot_state_publisher-3] Link right_track had 0 children
[robot_state_publisher-3] [INFO] [1657805371.091112900] [robot_state_publisher]: got segment back_left_wheel
[robot_state_publisher-3] [INFO] [1657805371.091357900] [robot_state_publisher]: got segment back_right_wheel
[robot_state_publisher-3] [INFO] [1657805371.091380100] [robot_state_publisher]: got segment camera
[robot_state_publisher-3] [INFO] [1657805371.091390700] [robot_state_publisher]: got segment chassis
[robot_state_publisher-3] [INFO] [1657805371.091399800] [robot_state_publisher]: got segment front_left_wheel
[robot_state_publisher-3] [INFO] [1657805371.091409000] [robot_state_publisher]: got segment front_right_wheel
[robot_state_publisher-3] [INFO] [1657805371.091417600] [robot_state_publisher]: got segment hokuyo
[robot_state_publisher-3] [INFO] [1657805371.091426500] [robot_state_publisher]: got segment imu_link
[robot_state_publisher-3] [INFO] [1657805371.091435300] [robot_state_publisher]: got segment left_track
[robot_state_publisher-3] [INFO] [1657805371.091444100] [robot_state_publisher]: got segment right_track
[robot_state_publisher-3] [INFO] [1657805371.091452200] [robot_state_publisher]: got segment robot_footprint
[spawn_entity.py-5] [INFO] [1657805371.534547900] [spawn_entity]: Spawn Entity started
[spawn_entity.py-5] [INFO] [1657805371.535182900] [spawn_entity]: Loading entity published on topic robot_description
[spawn_entity.py-5] [INFO] [1657805371.537033200] [spawn_entity]: Waiting for entity xml on robot_description
[spawn_entity.py-5] [INFO] [1657805371.541310400] [spawn_entity]: Waiting for service /spawn_entity, timeout = 30
[spawn_entity.py-5] [INFO] [1657805371.541762500] [spawn_entity]: Waiting for service /spawn_entity
[joint_state_publisher-4] [INFO] [1657805371.640828600] [joint_state_publisher]: Waiting for robot_description to be published on the robot_description topic...
[spawn_entity.py-5] [INFO] [1657805375.301664500] [spawn_entity]: Calling service /spawn_entity
[gzserver-1] [WARN] [1657805378.119790300] [rcl]: Found remap rule 'camera/image_raw:=/skidbot/camera/image'. This syntax is deprecated. Use '--ros-args --remap camera/image_raw:=/skidbot/camera/image' instead.
[gzserver-1] [WARN] [1657805378.120045400] [rcl]: Found remap rule 'camera/camera_info:=/skidbot/camera/image/camera_info'. This syntax is deprecated. Use '--ros-args --remap camera/camera_info:=/skidbot/camera/image/camera_info' instead.
[gzserver-1] [WARN] [1657805378.124065100] [rcl]: Found remap rule 'camera/image_raw:=/skidbot/camera/image'. This syntax is deprecated. Use '--ros-args --remap camera/image_raw:=/skidbot/camera/image' instead.
[gzserver-1] [WARN] [1657805378.124123200] [rcl]: Found remap rule 'camera/camera_info:=/skidbot/camera/image/camera_info'. This syntax is deprecated. Use '--ros-args --remap camera/camera_info:=/skidbot/camera/image/camera_info' instead.
[gzserver-1] [INFO] [1657805378.138004200] [skidbot.camera_controller]: Publishing camera info to [/skidbot/camera_sensor/camera_info]
[spawn_entity.py-5] [INFO] [1657805378.161223100] [spawn_entity]: Spawn status: SpawnEntity: Successfully spawned entity [fusionbot]
[gzserver-1] [WARN] [1657805378.179615500] [rcl]: Found remap rule '~/out:=scan'. This syntax is deprecated. Use '--ros-args --remap ~/out:=scan' instead.
[gzserver-1] [WARN] [1657805378.183462800] [rcl]: Found remap rule '~/out:=scan'. This syntax is deprecated. Use '--ros-args --remap ~/out:=scan' instead.
[gzserver-1] [INFO] [1657805378.199164100] [skidbot.imu.imu_plugin]: <initial_orientation_as_reference> is unset, using default value of false to comply with REP 145 (world as orientation reference)
[gzserver-1] [WARN] [1657805378.287742100] [rcl]: Found remap rule '/cmd_vel:=cmd_vel'. This syntax is deprecated. Use '--ros-args --remap /cmd_vel:=cmd_vel' instead.
[gzserver-1] [WARN] [1657805378.287814900] [rcl]: Found remap rule '/odom:=odom'. This syntax is deprecated. Use '--ros-args --remap /odom:=odom' instead.
[gzserver-1] [INFO] [1657805378.293983200] [skidbot.skid_steer_drive_controller]: Wheel pair 1 separation set to [0.400000m]
[gzserver-1] [INFO] [1657805378.294153600] [skidbot.skid_steer_drive_controller]: Wheel pair 2 separation set to [0.400000m]
[gzserver-1] [INFO] [1657805378.294249600] [skidbot.skid_steer_drive_controller]: Wheel pair 1 diameter set to [0.200000m]
[gzserver-1] [INFO] [1657805378.294273800] [skidbot.skid_steer_drive_controller]: Wheel pair 2 diameter set to [0.200000m]
[gzserver-1] [WARN] [1657805378.294370300] [rcl]: Found remap rule '/cmd_vel:=cmd_vel'. This syntax is deprecated. Use '--ros-args --remap /cmd_vel:=cmd_vel' instead.
[gzserver-1] [WARN] [1657805378.294385300] [rcl]: Found remap rule '/odom:=odom'. This syntax is deprecated. Use '--ros-args --remap /odom:=odom' instead.
[gzserver-1] [INFO] [1657805378.296394300] [skidbot.skid_steer_drive_controller]: Subscribed to [/skidbot/cmd_vel]
[gzserver-1] Error [Param.cc:449] Invalid argument. Unable to set value [world ] for key[odometry_source].
[gzserver-1] [INFO] [1657805378.298116300] [skidbot.skid_steer_drive_controller]: Advertise odometry on [/skidbot/odom]
[gzserver-1] [INFO] [1657805378.299764600] [skidbot.skid_steer_drive_controller]: Publishing odom transforms between [odom] and [robot_footprint]
[gzserver-1] [INFO] [1657805378.299833600] [skidbot.skid_steer_drive_controller]: Publishing wheel transforms between [robot_footprint], [front_left_wheel_joint] and [front_right_wheel_joint]
[gzserver-1] [INFO] [1657805378.299850400] [skidbot.skid_steer_drive_controller]: Publishing wheel transforms between [robot_footprint], [back_left_wheel_joint] and [back_right_wheel_joint]
[INFO] [spawn_entity.py-5]: process has finished cleanly [pid 1464]

밑에서 9번째줄에 [gzserver-1] Error [Param.cc:449] Invalid argument. Unable to set value [world ] for key[odometry_source].
라고 오류가 하나 떴습니다.

그리고 rqt_graph를 실행하였을 때
![오류12](https://user-images.githubusercontent.com/109124955/178998040-b5348b58-86e5-44ce-ba56-db881314a571.PNG)
와 같이 나와야 하지만

![오류11](https://user-images.githubusercontent.com/109124955/178998794-e93b7b4b-cea6-4bb3-819f-35f24290c4c5.PNG)
와 같이 t__clodck과 /skidbot/cmd_vel 그리고 /skidbot/teleop_twist_keyboard가 없습니다.

현재실행중인 노드는
(ROS 2 foxy) dnw0406@DESKTOP-MD97PB4:~/ros2_ws$
 ros2 node list
/gazebo
/joint_state_publisher
/robot_state_publisher
/rqt_gui_py_node_1792
/skidbot/camera_controller
/skidbot/gazebo_ros_head_hokuyo_controller
/skidbot/imu/imu_plugin
/skidbot/skid_steer_drive_controller
/teleop_twist_keyboard
로 떴습니다.

/teleop_twist_keyboard가 실행되고 있지만 rqt_graph에서는 나오지 않습니다.

프로그램은 ros2_ws에서 실행하였습니다.

t__clock은 무엇이고, 로봇 조종을 어떻게 하면 할 수 있을 지 알고 싶습니다.

**Comment by kimsooyoung** [comment]
자세한 설명 감사합니다!

우선, 로봇을 조종하기 위한 커멘드 라인을 입력한 터미널을 지금과 같이 보여주실 수 있을까요?

`ros2 node list`의 결과인 `/teleop_twist_keyboard`가 로봇을 조종하는 node인데 topic이 제대로 전달되지 않는 것 같습니다.
따라서 이와 같은 경우  `ros2 topic list`를 통해 topic 이름을 1차적으로 확인한 뒤,
잘못된 이름으로 인해 제대로 수신이 불가하였는지 디버깅이 가능할 것 같습니다. 

마지막으로 `t__clock` 가제보 시뮬레이션 상에서의 **시간을 publish하는 node**입니다. launch 파일 수정을 통해 gazebo만 실행시켜보신 뒤 다시 한 번 `rqt_graph`를 살펴보시면 어디서 문제가 생겼는지 더 정확하게 알 수 있을 것 같습니다.

**Reply by Jeoungji** [reply]
조종 터미널 내용입니다.

(ROS 2 foxy) dnw0406@DESKTOP-MD97PB4:~/ros2_ws$ ros2 run teleop_twist_keyboard teleop_twist_keyboard --ros-args -r   ns:=/skidbot

This node takes keypresses from the keyboard and publishes them
as Twist messages. It works best with a US keyboard layout.
---------------------------
Moving around:
   u    i    o
   j    k    l
   m    ,    .

For Holonomic mode (strafing), hold down the shift key:
---------------------------
   U    I    O
   J    K    L
   M    <    >

t : up (+z)
b : down (-z)

anything else : stop

q/z : increase/decrease max speeds by 10%
w/x : increase/decrease only linear speed by 10%
e/c : increase/decrease only angular speed by 10%

CTRL-C to quit

currently:      speed 0.5       turn 1.0

그리고 topic리스트입니다.

(ROS 2 foxy) dnw0406@DESKTOP-MD97PB4:~/ros2_ws$ ros2 topic list
/clock
/joint_states
/parameter_events
/performance_metrics
/robot_description
/rosout
/skidbot/camera_sensor/camera_info
/skidbot/camera_sensor/image_raw
/skidbot/cmd_vel
/skidbot/imu/data
/skidbot/odom
/skidbot/scan
/tf
/tf_static

**Reply by yoohr** [reply]
해결하셨으면 어떻게 해결하셨는지 알려주실수있으실까요?

**Reply by Jeoungji** [reply]
lauch.py에서 server, client cmd를 제외하고 주석 처리 한뒤 실행을 해보았는데, __clock이 뜨지 않아 ubuntu와 ros2를 삭제하고 다시 깔아보았습니다.

하지만 이전과 같이 __clock이 뜨질 않고 있습니다.
![오류21](https://user-images.githubusercontent.com/109124955/179355308-82023b53-f61f-46f9-8e03-d307884d1c5d.PNG)
![오류22](https://user-images.githubusercontent.com/109124955/179355314-7ef573a5-08cc-40f2-811f-5baacd671d0f.PNG)

실행 노드에도 clock이 뜨지 않습니다.
구글에 검색을 해보았지만 clock이 무엇인지는 나오지만 실행할 수 있는 방법을 찾지 못하였습니다.
혹시 clock 노드를 다운 받거나 실행시킬 다른 방안이 있을까요?
그리고 예제 프로그램의 노드와 토픽의 실행파일 위치를 알 수 있는 방법이 있을까요?

**Reply by Kim-JeongHan** [reply]
혹시 이 오류 기존의 ros1이 깔려있는 상황에서 gazebo를 한번 더 깔으셨나요?

**Reply by kimsooyoung** [reply]
해당 문제로 많은 분들께서 골치를 앓고 계신 것 같습니다.

사실 저 또한 처음 겪어보는 이슈인데 gazebo와 ros 사이 연동에서 문제가 생긴 것 같습니다. 

```
sudo apt install ros-foxy-gazebo-ros
sudo apt install ros-foxy-gazebo-ros-pkgs
```

위의 두 커멘드 라인을 실행하신 뒤 결과를 알려주시겠어요?

**Reply by Kim-JeongHan** [reply]
실행했음에도 결과차이가 없습니다

**Comment by Jeoungji** [comment]
노드 리스트: 
(ROS 2 foxy) dnw0406@DESKTOP-MD97PB4:~$ ros2 node list
/gazebo
/joint_state_publisher
/robot_state_publisher
/rqt_gui_py_node_1039
/skidbot/camera_controller
/skidbot/gazebo_ros_head_hokuyo_controller
/skidbot/imu/imu_plugin
/skidbot/skid_steer_drive_controller
/teleop_twist_keyboard

토픽 리스트 :
(ROS 2 foxy) dnw0406@DESKTOP-MD97PB4:~$ ros2 topic list
/clock
/cmd_vel
/joint_states
/parameter_events
/performance_metrics
/robot_description
/rosout
/skidbot/camera_sensor/camera_info
/skidbot/camera_sensor/image_raw
/skidbot/cmd_vel
/skidbot/imu/data
/skidbot/odom
/skidbot/scan
/tf
/tf_static

/teleop_twist_keyboard의 노드 정보 :
(ROS 2 foxy) dnw0406@DESKTOP-MD97PB4:~$ ros2 node info /teleop_twist_keyboard
/teleop_twist_keyboard
  Subscribers:

  Publishers:
    /cmd_vel: geometry_msgs/msg/Twist
    /parameter_events: rcl_interfaces/msg/ParameterEvent
    /rosout: rcl_interfaces/msg/Log
  Service Servers:
    /teleop_twist_keyboard/describe_parameters: rcl_interfaces/srv/DescribeParameters
    /teleop_twist_keyboard/get_parameter_types: rcl_interfaces/srv/GetParameterTypes
    /teleop_twist_keyboard/get_parameters: rcl_interfaces/srv/GetParameters
    /teleop_twist_keyboard/list_parameters: rcl_interfaces/srv/ListParameters
    /teleop_twist_keyboard/set_parameters: rcl_interfaces/srv/SetParameters
    /teleop_twist_keyboard/set_parameters_atomically: rcl_interfaces/srv/SetParametersAtomically
  Service Clients:

  Action Servers:

  Action Clients:

![335](https://user-images.githubusercontent.com/109124955/179573711-4927085a-40ff-4d6c-8af2-fa4d05e4b41b.PNG)

/teleop_twist_keyboard가 /skidbot안에 들어가 있지 않아 /skidbot/cmd_vel로 호출한 것이 아닌 /cmd_vel로 호출하였습니다. 이유는 모르겠지만 /teleop_twist_keyboard가 /skidbot안에 있지 않아 생긴 오류라고 생각합니다. 하지만 /teleop_twist_keyboard가 어디에 있는 파일인지를 찾지 못하여 노드를 수정할 수 는 없었습니다.


**skidbot의 위치를 알기 위해 인터넷에 검색한 결과 gazebo에서 airobo_gazebo와 똑같은 예제를 다루는 블로그( https://blog.naver.com/hanjiho97/222759661333 )를 찾았습니다.
블로그에서와 달리 설치하지 않은 파일들을 설치해보았습니다.**

**sudo apt install ros-foxy-image-view
sudo apt install ros-foxy-navigation2 ros-foxy-nav2-bringup** 

설치 결과
노드 리스트 :
(ROS 2 foxy) dnw0406@DESKTOP-MD97PB4:~$ ros2 node list
/gazebo
/joint_state_publisher
/robot_state_publisher
/skidbot/camera_controller
/skidbot/gazebo_ros_head_hokuyo_controller
/skidbot/imu/imu_plugin
/skidbot/skid_steer_drive_controller
/skidbot/teleop_twist_keyboard

토픽 리스트 :
(ROS 2 foxy) dnw0406@DESKTOP-MD97PB4:~$ ros2 topic list
/clock
/joint_states
/parameter_events
/performance_metrics
/robot_description
/rosout
/skidbot/camera_sensor/camera_info
/skidbot/camera_sensor/image_raw
/skidbot/cmd_vel
/skidbot/imu/data
/skidbot/odom
/skidbot/scan
/tf
/tf_static

![오류없음](https://user-images.githubusercontent.com/109124955/179575552-3f2ad3f6-6d1f-4131-b55a-31c6ce8ffc28.PNG)

![dhfbdqtdma2](https://user-images.githubusercontent.com/109124955/179575644-5f0b8b61-1e06-451e-ad8f-c6c66af44f74.PNG)

성공적으로 조종이 가능해졌습니다.

하지만 궁금한 것이 있습니다.
강의에서는 /gazebo와 skidbot을 연결하주는 것이 t__clock노드로 나와있습니다. 하지만 저는 /clock이라는 토픽이 이를 대신하고 있습니다. 둘은 어떤 차이가 있는 것이고, 왜 같은 결과가 나오는 것인가요?
그리고 skidbot에 관한 파일들 중 실행되고 있는 node와 동일한 이름의 파일은 없었습니다. 대신 skidbot이라는 이름의 urdf,riviz 등의 파일은 존재하였습니다. skidbot 관련 프로그램들은 /skidbot안에 모두 묶여 저장되어있는 것인가요?

**Reply by kimsooyoung** [reply]
해결되셨다니 다행입니다. 다만 image-view와 nav2는 이번 예시와 전혀 상관 없는 것이라 직접적인 관련이 있기보다는 설치 시 꼬여있는 것이 있던 것 같습니다.

1. t__clock과 clock 모두 gazebo에서의 simulation time을 흐르도록 하는 node입니다. 다만 t__ 라는 접두사가 붙는 것은 일종의 관습이며, 실제 로봇에서의 시간이 clock이라는 동일한 이름을 가질 수 있어 이름의 중복을 방지하기 위해 __ 를 많이 사용합니다. 다만 t__가 붙고 붙지 않고의 차이가 어디서 발생하는지는 저 또한 여러 환경에서 실험하여 살펴보겠습니다.


2. /skidbot을 앞에 붙이는 것을 namespace라고 합니다. /cmd_vel을 사용하는 여러 로봇 프로그램이 있을 수 있습니다. 일례로, turtlesim과 우리 gazebo예시를 둘 다 실행하고 /cmd_vel로 조종 신호를 주면 원치 않게 두 로봇 모두 움직이는 경우가 발생할 수 있습니다. 이러한 이유로 topic이름 앞에 추가적인 이름을 붙여주는 것을 namespace라고 합니다.
